prov/rxm/src/rxm_cq.c

```diff
@@ -759,6 +761,237 @@ static int rxm_handle_remote_write(struct rxm_ep *rxm_ep,
+
+	if (resp_len < rxm_ep->msg_info->tx_attr->inject_size) {
+		ret = fi_inject(rx_buf->conn->msg_ep, &resp_buf->pkt,
+				resp_len, 0);
+		if (OFI_LIKELY(!ret))
+			rxm_tx_buf_release(rxm_ep, RXM_BUF_POOL_TX_ATOMIC,
+					   resp_buf);
+	} else {
+		ret = rxm_atomic_send_respmsg(rxm_ep, rx_buf->conn, resp_buf,
+					      resp_len);
+	}
+	if (OFI_UNLIKELY(ret)) {
+		FI_WARN(&rxm_prov, FI_LOG_CQ,
+			"Unable to send Atomic Response\n");
+		if (OFI_LIKELY(ret == -FI_EAGAIN)) {
+			def_tx_entry =
+				rxm_ep_alloc_deferred_tx_entry(rxm_ep,
+						rx_buf->conn,
+						RXM_DEFERRED_TX_ATOMIC_RESP);
+			if (OFI_UNLIKELY(!def_tx_entry)) {
+				FI_WARN(&rxm_prov, FI_LOG_CQ,
+					"Unable to allocate deferred Atomic "
+					"Response\n");
+				return -FI_ENOMEM;
+			}
+
+			def_tx_entry->atomic_resp.tx_buf = resp_buf;
+			def_tx_entry->atomic_resp.len = resp_len;
+			rxm_ep_enqueue_deferred_tx_queue(def_tx_entry);
+			ret = 0;
+		}
+	}
+	rxm_enqueue_rx_buf_for_repost_check(rx_buf);
+
+	return ret;
+}
+
+static inline void rxm_do_atomic(struct rxm_pkt *pkt, void *dst, void *src,
+				 void *cmp, void *res, size_t count,
+				 enum fi_datatype datatype, enum fi_op op)
+{
+	switch (pkt->hdr.op) {
+	case ofi_op_atomic:
+		ofi_atomic_write_handlers[op][datatype](dst, src, count);
+		break;
+	case ofi_op_atomic_fetch:
+		ofi_atomic_readwrite_handlers[op][datatype](dst, src, res,
+							    count);
+		break;
+	case ofi_op_atomic_compare:
+		ofi_atomic_swap_handlers[op - OFI_SWAP_OP_START][datatype](dst,
+						src, cmp, res, count);
+		break;
+	default:
+		/* Validated prior to calling function */
+		break;
+	}
+}
+
+static inline ssize_t rxm_handle_atomic_req(struct rxm_ep *rxm_ep,
+					    struct rxm_rx_buf *rx_buf)
+{
+	struct rxm_atomic_hdr *req_hdr =
+			(struct rxm_atomic_hdr *) rx_buf->pkt.data;
+	enum fi_datatype datatype = rx_buf->pkt.hdr.atomic.datatype;
+	enum fi_op atomic_op = rx_buf->pkt.hdr.atomic.op;
+	size_t datatype_sz = ofi_datatype_size(datatype);
+	size_t len;
+	ssize_t result_len;
+	uint64_t offset;
+	int i;
+	int ret = 0;
+	struct rxm_tx_atomic_buf *resp_buf;
+	struct rxm_atomic_resp_hdr *resp_hdr;
+	struct rxm_domain *domain = container_of(rxm_ep->util_ep.domain,
+					 struct rxm_domain, util_domain);
+
+	assert(!(rx_buf->comp_flags &
+		 ~(FI_RECV | FI_RECV | FI_REMOTE_CQ_DATA)));
+	assert(rx_buf->pkt.hdr.op == ofi_op_atomic ||
+	       rx_buf->pkt.hdr.op == ofi_op_atomic_fetch ||
+	       rx_buf->pkt.hdr.op == ofi_op_atomic_compare);
+
+	if (rx_buf->ep->srx_ctx)
+		rx_buf->conn = rxm_key2conn(rx_buf->ep,
+					    rx_buf->pkt.ctrl_hdr.conn_id);
+	if (OFI_UNLIKELY(!rx_buf->conn))
+		return -FI_EOTHER;
+
+	resp_buf = (struct rxm_tx_atomic_buf *) rxm_tx_buf_get(rxm_ep,
+							RXM_BUF_POOL_TX_ATOMIC);
+	if (OFI_UNLIKELY(!resp_buf)) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_DATA,
+			"Unable to allocate from Atomic buffer pool\n");
+		/* TODO: Should this be -FI_ENOMEM - how does it get
+		 * processed again */
+		return -FI_EAGAIN;
```

Comment:
@shefty - Let me know if you want me to move forward with changes to be able to return NAKs in these two cases now, or if I should wait on that for now.

Reply:
fi_inject won't help, since that could return EAGAIN.  Can we swap the rx_buf->conn check, so that at least that flow sends a NAK?

I think the tx_buf pools grow dynamically, so we're probably as good as we can be here.

Reply:
If the mapping of the conn_id in packet header to a connection fails, I don't think it is possible to return a NAK (i.e. that mapping is our link back to the associated MSG_EP/connection). The other protocols like either issue an assert() or the check above, on the target side of request.